File: examples/weather_cli/impl_cpp/weather_cli.cc
// Reference C++ implementation of weather-cli.
//
// This file is the *legacy* artifact in the worked example: the
// archaeologist distills it into ../system.dx, the architect refines that
// spec, and the implementer regenerates the system in another language
// (see ../impl_python/) without ever reading this file.
//
// Network calls and JSON parsing are mocked for brevity. The structural
// contract surface (env-var handling, cache TTL by mtime, exit codes,
// stdout/stderr split) is intact and matches the contracts in
// ../system.dx.

#include <chrono>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <string>
#include <sys/stat.h>

namespace {

constexpr int kCacheTtlSeconds = 600;

std::string CacheFilePath() {
  const char* home = std::getenv("HOME");
  if (home == nullptr) {
    // Fall back to the working directory rather than crashing; the
    // contracts do not pin a path.
    return ".weather_cache.json";
  }
  return std::string(home) + "/.weather_cache.json";
}

bool IsCacheValid(const std::string& path) {
  struct stat result;
  if (stat(path.c_str(), &result) != 0) {
    return false;
  }
  const auto now = std::chrono::system_clock::now();
  const auto cache_time =
      std::chrono::system_clock::from_time_t(result.st_mtime);
  const auto duration =
      std::chrono::duration_cast<std::chrono::seconds>(now - cache_time);
  return duration.count() < kCacheTtlSeconds;
}

// Simulated upstream call. A real implementation would issue an HTTP
// request to the OpenMeteo (or equivalent) endpoint.
std::string FetchFromUpstream(const std::string& zip,
                              const std::string& /*api_key*/) {
  return std::string("{\"zip\": \"") + zip +
         "\", \"temp\": \"72F\", \"condition\": \"Sunny\"}";
}

}  // namespace

int main(int argc, char* argv[]) {
  if (argc < 2) {
    std::cerr << "Usage: weather_cli <zipcode> [--json]\n";
    return 1;
  }

  const std::string zip = argv[1];
  const bool json_output =
      (argc == 3 && std::string(argv[2]) == "--json");

  const char* api_key_env = std::getenv("WEATHER_API_KEY");
  if (api_key_env == nullptr || api_key_env[0] == '\0') {
    std::cerr << "Error: WEATHER_API_KEY environment variable not set.\n";
    return 1;
  }

  const std::string cache_path = CacheFilePath();
  std::string weather_data;

  if (IsCacheValid(cache_path)) {
    std::ifstream cache(cache_path);
    std::getline(cache, weather_data);
  }

  if (weather_data.empty()) {
    weather_data = FetchFromUpstream(zip, api_key_env);
    std::ofstream cache(cache_path);
    cache << weather_data;
  }

  if (json_output) {
    std::cout << weather_data << "\n";
  } else {
    // Human-readable summary; exact phrasing is unconstrained.
    std::cout << "Weather for " << zip << ": 72F, Sunny\n";
  }

  return 0;
}

File: examples/weather_cli/impl_python/weather_cli.py
#!/usr/bin/env python3
"""Reference Python implementation of weather-cli.

This file is the *agentic synthesis* artifact in the worked example: an
implementer agent reads only ../system.dx (never ../impl_cpp/) and
produces this file. The two implementations should be observably
indistinguishable from the perspective of the contracts in system.dx.

Network calls are mocked so the example runs offline. The contract
surface (env-var handling, cache TTL, exit codes, stdout/stderr split,
JSON output mode) is real.
"""

from __future__ import annotations

import argparse
import json
import os
import sys
import time
from pathlib import Path
from typing import Any

# Pinned by assumptions.cache.location in system.dx.
CACHE_FILE = Path.home() / ".weather_cache.json"

# Pinned by assumptions.network.provider in system.dx. Used only as a
# label in the JSON output; no real network call is issued.
UPSTREAM_PROVIDER = "OpenMeteo"

# Encoded directly in the perf_no_redundant_fetch invariant.
CACHE_TTL_SECONDS = 600


def is_cache_valid(path: Path = CACHE_FILE) -> bool:
    """Return True if the cache file exists and is younger than the TTL."""
    if not path.exists():
        return False
    age = time.time() - path.stat().st_mtime
    return age < CACHE_TTL_SECONDS


def fetch_weather(zip_code: str, _api_key: str) -> dict[str, Any]:
    """Simulated upstream fetch.

    A real implementation would issue an HTTP GET to the upstream
    provider here. The shape of the returned dict is what flows out
    through `--json`, so it is shaped to satisfy the
    `emits_json_with_flag` contract.
    """
    return {
        "zip": zip_code,
        "temp": "72F",
        "condition": "Sunny",
        "provider": UPSTREAM_PROVIDER,
    }


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(
        prog="weather_cli",
        description="Fetch and display current weather for a US zip code.",
    )
    parser.add_argument("zipcode", help="US zip code")
    parser.add_argument(
        "--json",
        dest="as_json",
        action="store_true",
        help="emit a single JSON object on stdout instead of a human summary",
    )
    args = parser.parse_args(argv)

    api_key = os.environ.get("WEATHER_API_KEY")
    if not api_key:
        print(
            "Error: WEATHER_API_KEY environment variable not set.",
            file=sys.stderr,
        )
        return 1

    weather: dict[str, Any] | None = None
    if is_cache_valid():
        try:
            with CACHE_FILE.open("r", encoding="utf-8") as f:
                weather = json.load(f)
        except (OSError, json.JSONDecodeError):
            weather = None  # Treat a corrupt cache as a miss.

    if weather is None:
        weather = fetch_weather(args.zipcode, api_key)
        try:
            with CACHE_FILE.open("w", encoding="utf-8") as f:
                json.dump(weather, f)
        except OSError:
            # Cache write failures are non-fatal: the user still gets
            # a fresh result this run.
            pass

    if args.as_json:
        print(json.dumps(weather))
    else:
        print(
            f"Weather for {weather['zip']}: {weather['temp']}, {weather['condition']}"
        )
    return 0


if __name__ == "__main__":
    sys.exit(main())
